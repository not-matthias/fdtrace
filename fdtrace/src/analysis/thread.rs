use super::file::{FileInfo, FileSession};
use crate::{
    analysis::{file::FileEvent, utils},
    syscall::{fd_t, tid_t, RawSyscall, Syscall},
};
use itertools::Itertools;
use std::collections::HashMap;

#[derive(Debug)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct ThreadAnalysis {
    tid: tid_t,
    files: HashMap<String, FileInfo>,
}

impl ThreadAnalysis {
    pub fn new(tid: tid_t, syscalls: &[Syscall]) -> Self {
        log::info!("Thread {tid} got {} syscalls", syscalls.len());

        let mut files = HashMap::new();

        // All the current sessions. A new session is created when the file is opened,
        // and is removed from this list and added to `files` when the file is closed.
        //
        let mut cur_sessions = {
            let mut map = HashMap::new();

            // Add the default sessions for stdin, stdout, and stderr
            map.insert(0, FileSession::new("/dev/stdin"));
            map.insert(1, FileSession::new("/dev/stdout"));
            map.insert(2, FileSession::new("/dev/stderr"));

            map
        };

        let mut iter = syscalls.iter().multipeek();
        while let Some(call) = iter.next() {
            assert_eq!(tid, call.tid);

            match &call.raw {
                RawSyscall::OpenAt { path, .. } | RawSyscall::Open { path, .. } => {
                    let Some(RawSyscall::OpenExit { ret } | RawSyscall::OpenAtExit { ret }) =
                        iter.peek().map(|s| &s.raw)
                    else {
                        // Sometimes we don't directly get a exit
                        //
                        // Example:
                        // openat;-100;/target/...;524288
                        // openat;-1059213040;;-1059213040
                        //
                        log::warn!("Syscall not followed by exit: {call:?}");
                        continue;
                    };

                    if *ret != -1 {
                        cur_sessions.insert(
                            *ret as fd_t,
                            FileSession {
                                path: path.clone(),
                                open_ts: call.ts,
                                ..Default::default()
                            },
                        );
                        log::debug!("Created a new session for {path}");
                    }
                }
                RawSyscall::Read { fd: read_fd, .. } => {
                    let cur_session = cur_sessions
                        .get_mut(read_fd)
                        .unwrap_or_else(|| panic!("Can't read without open: {call:?}"));

                    let (end_ts, read) = match iter.peek().map(|s| (s.ts, &s.raw)) {
                        Some((end_ts, RawSyscall::ReadExit { read })) => (end_ts, read),
                        _ => {
                            // In some cases, the exit syscall is delayed by one
                            // syscall. We can check the next syscall to see if it's
                            // the exit syscall.
                            //
                            // 0;92062;92062;read;5;4096
                            // 1;92062;92062;read;-1050723616;18446631905292925664
                            // 3;92062;92062;read_exit;4096
                            // 4;92062;92062;read_exit;-1050734920
                            //
                            let Some((end_ts, RawSyscall::ReadExit { read })) =
                                iter.peek().map(|s| (s.ts, &s.raw))
                            else {
                                panic!("Read syscall not followed by read exit: {call:?}")
                            };

                            (end_ts, read)
                        }
                    };

                    // Only record if read was successful
                    // - 0 = EOF
                    // - -1 = error
                    //
                    if read > &0 {
                        cur_session.events.push(FileEvent::Read {
                            bytes: *read as usize,
                            start_ts: call.ts,
                            end_ts,
                        });
                        log::debug!("Read {read} bytes from {}", cur_session.path);
                    }
                }
                RawSyscall::Write { fd, .. } => {
                    let cur_session = cur_sessions
                        .get_mut(fd)
                        .unwrap_or_else(|| panic!("Can't write without open: {call:?}"));

                    let Some((end_ts, RawSyscall::WriteExit { written })) =
                        iter.peek().map(|s| (s.ts, &s.raw))
                    else {
                        panic!("Write syscall not followed by write exit")
                    };

                    // Only record if write was successful.
                    // - 0 = nothing was written
                    // - -1 = error
                    //
                    if written > &-0 {
                        cur_session.events.push(FileEvent::Write {
                            bytes: *written as usize,
                            start_ts: call.ts,
                            end_ts,
                        });
                        log::debug!("Wrote {written} bytes to {{session.path}}");
                    }
                }
                RawSyscall::Close { fd: close_fd } => {
                    let mut cur_session = cur_sessions
                        .remove(close_fd)
                        .unwrap_or_else(|| panic!("Close without open: {call:?}"));
                    log::debug!("Closed {{cur_session.path}}");

                    let file_info = files
                        .entry(cur_session.path.clone())
                        .or_insert(FileInfo::default());
                    cur_session.close_ts = call.ts;
                    file_info.sessions.push(cur_session);
                }

                _ => {}
            }
        }

        // Close the stdin, stdout, and stderr sessions
        //
        #[cfg(feature = "trace-stdfd")]
        for fd in 0..3 {
            let mut cur_session = cur_sessions.remove(&fd).unwrap();
            cur_session.open_ts = syscalls.first().unwrap().ts;
            cur_session.close_ts = syscalls.last().unwrap().ts;

            let file_info = files
                .entry(cur_session.path.clone())
                .or_insert(Default::default());
            file_info.sessions.push(cur_session);
        }

        Self { tid, files }
    }

    pub fn print_result(&self) {
        use termimad::print_inline as mdprintln;

        mdprintln(&format!("\n# **Thread: {}**\n\n", self.tid));
        for (path, file_info) in &self.files {
            mdprintln(&format!("\n## File: **{}**\n\n", path));
            println!("Opened: {} times", file_info.sessions.len());

            let total_duration = utils::ns_to_ms(
                file_info.sessions.iter().map(|s| s.duration()).sum::<u64>() as f64,
            );
            let avg_duration = total_duration / file_info.sessions.len() as f64;
            mdprintln(&format!("Total duration: {:.2} ms\n", total_duration));
            mdprintln(&format!("Avg session duration: {:.2} ms\n", avg_duration));

            // Sessions
            //
            println!();
            for (i, session) in file_info.sessions.iter().enumerate() {
                mdprintln(&format!(
                    "**Session {}** was open for {:.2}ms (idle for {:.2}ms)\n",
                    i + 1,
                    session.duration_ms(),
                    session.idle_time_ms()
                ));

                for (j, event) in session.events.iter().enumerate() {
                    match event {
                        FileEvent::Read { bytes, .. } => {
                            mdprintln(&format!("- **Event {}**: Read {} bytes\n", j + 1, bytes));
                        }
                        FileEvent::Write { bytes, .. } => {
                            mdprintln(&format!("- **Event {}**: Write {} bytes\n", j + 1, bytes));
                        }
                    }
                }
            }

            // IO Statistics
            //
            let (total_read, total_write) = file_info.total_bytes();
            let (avg_read, avg_write) = file_info.avg_size().unwrap_or_default();
            let (max_read, max_write) = file_info.max_size();

            let mut io_table = comfy_table::Table::new();
            io_table
                .set_header(vec!["", "Read", "Write"])
                .add_row(vec![
                    "Total",
                    &total_read.to_string(),
                    &total_write.to_string(),
                ])
                .add_row(vec![
                    "Average",
                    &avg_read.to_string(),
                    &avg_write.to_string(),
                ])
                .add_row(vec!["Max", &max_read.to_string(), &max_write.to_string()]);

            // mdprintln("\n### **IO Statistics**:\n\n");
            println!();
            println!("{io_table}");
        }
    }
}
