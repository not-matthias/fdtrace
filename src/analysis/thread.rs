use super::file::{FileInfo, FileSession};
use crate::{
    analysis::{file::FileEvent, utils},
    syscall::{fd_t, tid_t, RawSyscall, Syscall},
};
use itertools::Itertools;
use std::collections::HashMap;

#[derive(Debug)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct ThreadAnalysis {
    tid: tid_t,
    files: HashMap<String, FileInfo>,
}

impl ThreadAnalysis {
    pub fn new(tid: tid_t, syscalls: &[Syscall]) -> Self {
        log::info!("Thread {tid} got {} syscalls", syscalls.len());

        let mut files = HashMap::new();

        // All the current sessions. A new session is created when the file is opened,
        // and is removed from this list and added to `files` when the file is closed.
        //
        let mut cur_sessions = {
            let mut map = HashMap::new();

            // Add the default sessions for stdin, stdout, and stderr
            map.insert(0, FileSession::new("/dev/stdin"));
            map.insert(1, FileSession::new("/dev/stdout"));
            map.insert(2, FileSession::new("/dev/stderr"));

            map
        };

        let mut iter = syscalls.iter().multipeek();
        while let Some(call) = iter.next() {
            assert_eq!(tid, call.tid);

            match &call.raw {
                RawSyscall::OpenAt { path, .. } | RawSyscall::Open { path, .. } => {
                    let Some(RawSyscall::OpenExit { ret } | RawSyscall::OpenAtExit { ret }) =
                        iter.peek().map(|s| &s.raw)
                    else {
                        log::warn!("Syscall not followed by exit: {call:?}");
                        continue;
                    };

                    if *ret != -1 {
                        cur_sessions.insert(
                            *ret as fd_t,
                            FileSession {
                                path: path.clone(),
                                open_ts: call.ts,
                                ..Default::default()
                            },
                        );
                        log::debug!("Created a new session for {path}");
                    }
                }

                RawSyscall::Read { fd, .. } | RawSyscall::Write { fd, .. } => {
                    let is_read = matches!(call.raw, RawSyscall::Read { .. });

                    let Some(cur_session) = cur_sessions.get_mut(fd) else {
                        log::warn!("RW without open: {call:?}");
                        continue;
                    };

                    macro_rules! peek_peek_exit {
                        ($name:path) => {{
                            let Some((end_ts, $name { count })) = iter
                                .peek()
                                .map(|s| (s.ts, &s.raw))
                                .or_else(|| iter.peek().map(|s| (s.ts, &s.raw)))
                            else {
                                // FIXME: We potentially lost a read event here. But it's still
                                // better to continue instead of panicking.
                                log::warn!("Read syscall not followed by read exit: {call:?}");
                                continue;
                            };

                            (end_ts, count)
                        }};
                    }

                    let (end_ts, count) = if is_read {
                        peek_peek_exit!(RawSyscall::ReadExit)
                    } else {
                        peek_peek_exit!(RawSyscall::WriteExit)
                    };

                    // Only record if read was successful
                    // - 0 = EOF
                    // - -1 = error
                    //
                    if count > &0 {
                        if is_read {
                            cur_session.events.push(FileEvent::Read {
                                bytes: *count as usize,
                                start_ts: call.ts,
                                end_ts,
                            });
                        } else {
                            cur_session.events.push(FileEvent::Write {
                                bytes: *count as usize,
                                start_ts: call.ts,
                                end_ts,
                            });
                        }
                    }
                }

                RawSyscall::Close { fd: close_fd } => {
                    let Some(mut cur_session) = cur_sessions.remove(close_fd) else {
                        // FIXME: More syscalls need to be traced to also catch this.
                        log::warn!("Close without open: {call:?}");
                        continue;
                    };
                    cur_session.close_ts = call.ts;
                    log::debug!("Closed {}", cur_session.path);

                    let file_info = files
                        .entry(cur_session.path.clone())
                        .or_insert(FileInfo::default());
                    file_info.sessions.push(cur_session);
                }

                _ => {}
            }
        }

        // Close the stdin, stdout, and stderr sessions
        //
        #[cfg(feature = "trace-stdfd")]
        for fd in 0..3 {
            let mut cur_session = cur_sessions.remove(&fd).unwrap();
            cur_session.open_ts = syscalls.first().unwrap().ts;
            cur_session.close_ts = syscalls.last().unwrap().ts;

            let file_info = files
                .entry(cur_session.path.clone())
                .or_insert(Default::default());
            file_info.sessions.push(cur_session);
        }

        Self { tid, files }
    }

    pub fn print_result(&self) {
        use termimad::print_inline as mdprintln;

        mdprintln(&format!("\n# **Thread: {}**\n\n", self.tid));
        for (path, file_info) in &self.files {
            mdprintln(&format!("\n## File: **{}**\n\n", path));
            println!("Opened: {} times", file_info.sessions.len());

            let total_duration = utils::ns_to_ms(
                file_info.sessions.iter().map(|s| s.duration()).sum::<u64>() as f64,
            );
            let avg_duration = total_duration / file_info.sessions.len() as f64;
            mdprintln(&format!("Total duration: {:.2} ms\n", total_duration));
            mdprintln(&format!("Avg session duration: {:.2} ms\n", avg_duration));

            // Sessions
            //
            println!();
            for (i, session) in file_info.sessions.iter().enumerate() {
                mdprintln(&format!(
                    "**Session {}** was open for {:.2}ms (idle for {:.2}ms)\n",
                    i + 1,
                    session.duration_ms(),
                    session.idle_time_ms()
                ));

                for (j, event) in session.events.iter().enumerate() {
                    match event {
                        FileEvent::Read { bytes, .. } => {
                            mdprintln(&format!("- **Event {}**: Read {} bytes\n", j + 1, bytes));
                        }
                        FileEvent::Write { bytes, .. } => {
                            mdprintln(&format!("- **Event {}**: Write {} bytes\n", j + 1, bytes));
                        }
                    }
                }
            }

            // IO Statistics
            //
            let (total_read, total_write) = file_info.total_bytes();
            let (avg_read, avg_write) = file_info.avg_size().unwrap_or_default();
            let (max_read, max_write) = file_info.max_size();

            let mut io_table = comfy_table::Table::new();
            io_table
                .set_header(vec!["", "Read", "Write"])
                .add_row(vec![
                    "Total",
                    &total_read.to_string(),
                    &total_write.to_string(),
                ])
                .add_row(vec![
                    "Average",
                    &avg_read.to_string(),
                    &avg_write.to_string(),
                ])
                .add_row(vec!["Max", &max_read.to_string(), &max_write.to_string()]);

            // mdprintln("\n### **IO Statistics**:\n\n");
            println!();
            println!("{io_table}");
        }
    }
}
