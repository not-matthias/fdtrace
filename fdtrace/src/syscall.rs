use std::ops::{Deref, DerefMut};

#[allow(non_camel_case_types)]
pub type pid_t = i32;
#[allow(non_camel_case_types)]
pub type tid_t = i32;

#[allow(non_camel_case_types)]
pub type fd_t = u64;

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Syscall {
    pub ts: u64,
    pub pid: pid_t,
    pub tid: tid_t,
    pub raw: RawSyscall,
}

impl Syscall {
    pub fn from_parts(data: &str) -> Option<Self> {
        let parts = data.split(";");
        Self::from_parts_iter(parts)
    }

    pub fn from_parts_iter<'a>(mut parts: impl Iterator<Item = &'a str>) -> Option<Self> {
        Some(Self {
            ts: parts.next()?.parse().ok()?,
            pid: parts.next()?.parse().ok()?,
            tid: parts.next()?.parse().ok()?,
            raw: RawSyscall::from_parts_iter(parts)?,
        })
    }
}

impl Deref for Syscall {
    type Target = RawSyscall;

    fn deref(&self) -> &Self::Target {
        &self.raw
    }
}

impl DerefMut for Syscall {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.raw
    }
}

/// # Covered syscalls
///
/// - File creation and opening: open, openat.
/// - File descriptor operations: close, read, write.
/// - File removal: unlink, unlinkat, rmdir.
/// - File renaming: rename, renameat.
/// - Directory creation: mkdir, mkdirat.
#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(serde::Serialize))]
#[rustfmt::skip]
pub enum RawSyscall {
    Execve { path: String },

    Open { path: String, flags: u64, mode: u64 },
    OpenExit { ret: i64 },

    OpenAt { dirfd: fd_t, path: String, flags: u64 },
    OpenAtExit { ret: i64 },

    Close { fd: fd_t },
    CloseExit { ret: i64 },

    Read { fd: fd_t, count: usize },
    ReadExit { read: i64 },

    Write { fd: fd_t, count: usize },
    WriteExit { written: i64 },
}

impl RawSyscall {
    pub fn from_parts(data: &str) -> Option<Self> {
        let parts = data.split(";");
        Self::from_parts_iter(parts)
    }

    /// Parses the ';' separated syscall (e.g. `read;42;42`)
    pub fn from_parts_iter<'a>(mut parts: impl Iterator<Item = &'a str>) -> Option<Self> {
        macro_rules! parse_syscall {
            ($syscall:ident, $($field:ident),*) => {
                Some(RawSyscall::$syscall {
                    $($field: parts.next()?.parse().ok()?,)*
                })
            };
        }

        match parts.next()? {
            "execve" => parse_syscall!(Execve, path),

            "open" => parse_syscall!(Open, path, flags, mode),
            "open_exit" => parse_syscall!(OpenExit, ret),

            "openat" => parse_syscall!(OpenAt, dirfd, path, flags),
            "openat_exit" => parse_syscall!(OpenAtExit, ret),

            "close" => parse_syscall!(Close, fd),
            "close_exit" => parse_syscall!(CloseExit, ret),

            "read" => parse_syscall!(Read, fd, count),
            "read_exit" => parse_syscall!(ReadExit, read),

            "write" => parse_syscall!(Write, fd, count),
            "write_exit" => parse_syscall!(WriteExit, written),

            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tracer::BpfTracer;

    #[test]
    fn test_parse_threaded() {
        let raw_trace = include_str!("../data/threaded.txt");
        let syscalls = BpfTracer::parse_trace(&raw_trace).unwrap();

        insta::assert_json_snapshot!(syscalls);
    }

    #[test]
    fn test_parse_multisession() {
        let raw_trace = include_str!("../data/multisession.txt");
        let syscalls = BpfTracer::parse_trace(&raw_trace).unwrap();

        insta::assert_json_snapshot!(syscalls);
    }

    #[test]
    fn test_parse_read_exit() {
        let parts = "read_exit;832";
        let syscall = RawSyscall::from_parts(parts).unwrap();
        assert_eq!(syscall, RawSyscall::ReadExit { read: 832 });
    }

    #[test]
    fn test_parse_close() {
        let parts =
            "20429708185183;105898;105898;openat;18446631905284431216;;18446631905284431216";
        let syscall = Syscall::from_parts(parts).unwrap();
        assert_eq!(syscall.ts, 18836222727359);
        assert_eq!(syscall.pid, 99220);
        assert_eq!(syscall.tid, 99220);
        assert_eq!(
            syscall.raw,
            RawSyscall::Close {
                fd: 18446631905284438752
            }
        );
    }
}
