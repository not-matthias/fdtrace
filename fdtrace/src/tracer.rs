use crate::syscall::Syscall;
use std::path::Path;
use tempfile::NamedTempFile;

pub struct BpfTracer {
    syscalls: Vec<Syscall>,
}

impl BpfTracer {
    pub fn trace(program: &Path) -> anyhow::Result<Self> {
        let tmpfile = NamedTempFile::new()?;
        let cmd = std::process::Command::new("bpftrace")
            .arg("-c")
            .arg(program)
            .arg("-o")
            .arg(tmpfile.path())
            // TODO: Fix this path
            .arg("scripts/fdtrace.bt")
            .output()?;

        if !cmd.status.success() {
            let error = String::from_utf8_lossy(&cmd.stderr);
            return Err(anyhow::anyhow!("{error}"));
        }

        let output = std::fs::read_to_string(tmpfile)?;
        Self::parse_trace(&output)
    }

    pub fn parse_trace(trace: &str) -> anyhow::Result<Self> {
        let parse_pid = |line: &str| {
            line.split(';')
                .next()
                .and_then(|p: &str| p.parse::<u32>().ok())
                .unwrap_or_default()
        };

        let mut target_pid = None;
        let mut syscalls = Vec::new();
        for line in trace.lines() {
            let Some(target_pid) = target_pid else {
                // The output contains many other processes logs as well, which is not what we
                // want. We need to find the 'execve' syscall to find the process id of our
                // target process.
                //
                if line.contains("execve") {
                    target_pid = Some(parse_pid(line));
                }

                continue;
            };

            // After we have our target_pid, we can filter out all the other logs that
            // aren't related to this process.
            //
            if parse_pid(line) != target_pid {
                continue;
            }

            let parts = line.split(';').skip(1).collect::<Vec<_>>().join(";");

            if let Some(syscall) = Syscall::from_parts(&parts) {
                syscalls.push(syscall);
            } else {
                // TODO: Return as Result<_, _>
                log::error!("Failed to parse syscall: {}", parts);
            }
        }

        Ok(Self { syscalls })
    }

    /// Prints the syscalls to a file
    pub fn print_to_file(&self, file: &str) {
        use std::io::Write;

        let mut file = std::fs::File::create(file).unwrap();
        for syscall in &self.syscalls {
            writeln!(file, "{:?}", syscall).unwrap();
        }
    }

    /// Prints the syscalls to stdout
    pub fn debug_print(&self) {
        for syscall in &self.syscalls {
            println!("{:?}", syscall);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Run with: sudo cargo t
    #[test]
    fn test_trace_ls() {
        let tracer =
            BpfTracer::trace("/nix/store/sbxy42ph4gjlg567vaz1kihmgiwqa5dh-system-path/bin/ls")
                .unwrap();
        assert_ne!(tracer.syscalls.len(), 0);
    }
}
