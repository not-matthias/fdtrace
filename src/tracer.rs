use crate::syscall::{RawSyscall, Syscall};
use std::{io::Write, path::Path};
use tempfile::NamedTempFile;

pub struct BpfTracer {
    syscalls: Vec<Syscall>,
}

impl BpfTracer {
    pub fn trace(program: &Path) -> anyhow::Result<Self> {
        let script = {
            let mut file = NamedTempFile::new()?;
            writeln!(file, "{}", include_str!("../data/fdtrace.bt"))?;
            file
        };

        let tmpfile = NamedTempFile::new()?;
        let cmd = std::process::Command::new("bpftrace")
            .arg("-c")
            .arg(program)
            .arg("-o")
            .arg(tmpfile.path())
            .arg(script.path())
            .output()?;

        if !cmd.status.success() {
            let error = String::from_utf8_lossy(&cmd.stderr);
            return Err(anyhow::anyhow!("{error}"));
        }

        let output = std::fs::read_to_string(tmpfile)?;
        std::fs::write("raw_trace.txt", &output).unwrap();

        Ok(Self {
            syscalls: Self::parse_trace(&output)?,
        })
    }

    pub fn parse_trace(trace: &str) -> anyhow::Result<Vec<Syscall>> {
        let mut target_pid = None;

        let mut syscalls = Vec::new();
        for line in trace.lines().skip(1) {
            if line.starts_with("Lost ") {
                log::warn!("Lost events: {line}");
                continue;
            }

            let syscall =
                Syscall::from_parts(line).unwrap_or_else(|| panic!("Failed to parse: {line}"));

            // The output contains many other processes logs as well, which is not what we
            // want. We need to find the 'execve' syscall to find the process id of our
            // target process.
            //
            let Some(target_pid) = target_pid else {
                if let RawSyscall::Execve { path } = &syscall.raw {
                    log::info!("Target process: {:?}", path);
                    target_pid = Some(syscall.pid);
                }
                continue;
            };

            // After we have our target_pid, we can filter out all the other logs that
            // aren't related to this process.
            //
            if syscall.pid != target_pid {
                continue;
            }

            syscalls.push(syscall);
        }

        Ok(syscalls)
    }

    pub fn syscalls(&self) -> &[Syscall] {
        &self.syscalls
    }

    pub fn take_syscalls(self) -> Vec<Syscall> {
        self.syscalls
    }

    /// Prints the syscalls to a file
    pub fn print_to_file(&self, file: &str) {
        use std::io::Write;

        let mut file = std::fs::File::create(file).unwrap();
        for syscall in &self.syscalls {
            writeln!(file, "{:?}", syscall).unwrap();
        }
    }

    /// Prints the syscalls to stdout
    pub fn debug_print(&self) {
        for syscall in &self.syscalls {
            println!("{:?}", syscall);
        }
    }
}
