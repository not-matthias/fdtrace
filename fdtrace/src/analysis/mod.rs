use crate::syscall::{tid_t, Syscall};
use itertools::Itertools;
use std::collections::HashMap;
use thread::ThreadAnalysis;

pub mod file;
pub mod thread;
mod utils;

#[derive(Debug)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Analysis {
    threads: HashMap<tid_t, ThreadAnalysis>,
}

impl Analysis {
    pub fn new(syscalls: Vec<Syscall>) -> Self {
        let threads = syscalls
            .into_iter()
            .into_group_map_by(|s| s.tid)
            .into_iter()
            .map(|(tid, syscalls)| (tid, ThreadAnalysis::new(tid, &syscalls)))
            .collect();

        Self { threads }
    }

    pub fn print_result(&self) {
        for (_, thread) in &self.threads {
            thread.print_result();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tracer::BpfTracer;

    #[test]
    fn test_analyze_multisession() {
        let raw_trace = include_str!("../../data/multisession.txt");
        let syscalls = BpfTracer::parse_trace(raw_trace).unwrap();
        let analysis = Analysis::new(syscalls);

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!(analysis);
        });
    }

    #[test]
    fn test_analyze_threaded() {
        let raw_trace = include_str!("../../data/threaded.txt");
        let syscalls = BpfTracer::parse_trace(raw_trace).unwrap();
        let analysis = Analysis::new(syscalls);

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!(analysis);
        });
    }
}
