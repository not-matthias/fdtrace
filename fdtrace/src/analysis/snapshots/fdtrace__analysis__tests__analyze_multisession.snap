---
source: fdtrace/src/analysis/mod.rs
expression: analysis
---
{
  "threads": {
    "110089": {
      "tid": 110089,
      "files": {
        "/etc/hostname": {
          "sessions": [
            {
              "events": [
                {
                  "Read": {
                    "bytes": 7,
                    "start_ts": 21197750336537,
                    "end_ts": 21197750337055
                  }
                }
              ],
              "path": "/etc/hostname",
              "open_ts": 21197750315606,
              "close_ts": 21198750532596
            }
          ]
        },
        "/etc/hosts": {
          "sessions": [
            {
              "events": [
                {
                  "Read": {
                    "bytes": 62,
                    "start_ts": 21197750321291,
                    "end_ts": 21197750322687
                  }
                }
              ],
              "path": "/etc/hosts",
              "open_ts": 21197750307894,
              "close_ts": 21198750537823
            }
          ]
        },
        "/etc/passwd": {
          "sessions": [
            {
              "events": [
                {
                  "Read": {
                    "bytes": 4089,
                    "start_ts": 21197750331770,
                    "end_ts": 21197750332500
                  }
                }
              ],
              "path": "/etc/passwd",
              "open_ts": 21197750314022,
              "close_ts": 21198750537187
            }
          ]
        },
        "/nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/li..": {
          "sessions": [
            {
              "events": [
                {
                  "Read": {
                    "bytes": 832,
                    "start_ts": 21197750081926,
                    "end_ts": 21197750082594
                  }
                }
              ],
              "path": "/nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/li..",
              "open_ts": 21197750080494,
              "close_ts": 21197750106083
            }
          ]
        },
        "/nix/store/rxganm4ibf31qngal3j3psp20mak37yy-xgcc-13.2.0-libgcc/..": {
          "sessions": [
            {
              "events": [
                {
                  "Read": {
                    "bytes": 832,
                    "start_ts": 21197750052776,
                    "end_ts": 21197750054156
                  }
                }
              ],
              "path": "/nix/store/rxganm4ibf31qngal3j3psp20mak37yy-xgcc-13.2.0-libgcc/..",
              "open_ts": 21197750049773,
              "close_ts": 21197750073346
            }
          ]
        },
        "/proc/self/maps": {
          "sessions": [
            {
              "events": [
                {
                  "Read": {
                    "bytes": 1024,
                    "start_ts": 21197750267478,
                    "end_ts": 21197750274761
                  }
                },
                {
                  "Read": {
                    "bytes": 1024,
                    "start_ts": 21197750282905,
                    "end_ts": 21197750286355
                  }
                },
                {
                  "Read": {
                    "bytes": 1024,
                    "start_ts": 21197750288772,
                    "end_ts": 21197750291308
                  }
                },
                {
                  "Read": {
                    "bytes": 417,
                    "start_ts": 21197750293104,
                    "end_ts": 21197750294269
                  }
                }
              ],
              "path": "/proc/self/maps",
              "open_ts": 21197750247373,
              "close_ts": 21197750295559
            }
          ]
        }
      }
    }
  }
}
